distributed/shuffle/_worker_extension.py

```diff
@@ -192,6 +151,336 @@ def heartbeat(self) -> dict[str, Any]:
+    run_id:
+        A unique identifier of the specific execution of the shuffle this belongs to.
+    local_address:
+        The local address this Shuffle can be contacted by using `rpc`.
+    directory:
+        The scratch directory to buffer data in.
+    nthreads:
+        How many background threads to use for compute.
+    loop:
+        The event loop.
+    rpc:
+        A callable returning a PooledRPCCall to contact other Shuffle instances.
+        Typically a ConnectionPool.
+    scheduler:
+        A PooledRPCCall to to contact the scheduler.
+    memory_limiter_disk:
+    memory_limiter_comm:
+        A ``ResourceLimiter`` limiting the total amount of memory used in either
+        buffer.
+    """
+
+    def __init__(
+        self,
+        worker_for: dict[NIndex, str],
+        output_workers: set,
+        old: ChunkedAxes,
+        new: ChunkedAxes,
+        id: ShuffleId,
+        run_id: int,
+        local_address: str,
+        directory: str,
+        nthreads: int,
+        rpc: Callable[[str], PooledRPCCall],
+        scheduler: PooledRPCCall,
+        memory_limiter_disk: ResourceLimiter,
+        memory_limiter_comms: ResourceLimiter,
+    ):
+        from dask.array.rechunk import _old_to_new
+
+        super().__init__(
+            id=id,
+            run_id=run_id,
+            output_workers=output_workers,
+            local_address=local_address,
+            directory=directory,
+            nthreads=nthreads,
+            rpc=rpc,
+            scheduler=scheduler,
+            memory_limiter_comms=memory_limiter_comms,
+            memory_limiter_disk=memory_limiter_disk,
+            dump=dump_shards,
+            load=load_partition,
+        )
+        self.old = old
+        self.new = new
+        partitions_of = defaultdict(list)
+        for part, addr in worker_for.items():
+            partitions_of[addr].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = worker_for
+        self._slicing = rechunk_slicing(old, new)
+        self._old_to_new = _old_to_new(old, new)
+
+    async def _receive(self, data: list[tuple[tuple[NIndex, NIndex], bytes]]) -> None:
+        self.raise_if_closed()
+
+        filtered = []
+        for d in data:
+            if d[0] not in self.received:
+                filtered.append(d)
+                self.received.add(d[0])
+                self.total_recvd += sizeof(d)
+        del data
+        if not filtered:
+            return
+        try:
+            buffers = await self.offload(self._reserialize_buffers, filtered)
+            del filtered
+            await self._write_to_disk(buffers)
+        except Exception as e:
+            self._exception = e
+            raise
+
+    def _reserialize_buffers(
+        self, data: list[tuple[tuple[NIndex, NIndex], bytes]]
+    ) -> dict[NIndex, list[bytes]]:
+        result = defaultdict(list)
+        for d in data:
+            result[d[0][0]].append(d[1])
```

Comment:
why is only the first part of the index relevant here? IIUC the `d[0][0]` is selecting the first `NIndex`.

I might be missing some context about what the first and the second `NIndex` represents. Is this documented somewhere?

Reply:
It's documented a few lines below in the docstring for `_()`, but I've been thinking about adding a `dataclass` for this pondering whether it would be useful for readability. I guess the answer to that is: Yes.

Reply:
I've added `ArrayRechunkShardID`, I hope this makes things clearer now.